juju/juju
# state: check constraints when destroying storage

## Description of change

When detaching or destroying storage, ensure that
doing so does not violate the minimum storage
requirements of the owner (unit or application).
The minimum storage requirements are ignored when
the owner is dying, to allow storage to be cleanly
detached during unit destruction.

Also: rename DestroyStorageAttachment to
DetachStorage, to better match what we have with
volumes and filesystems. An AttachStorage method
will be added in a future PR.

## QA steps

1. export JUJU_DEV_FEATURE_FLAGS=lxd-storage
2. juju bootstrap localhost
3. juju deploy /path/to/storagetest --storage fs=lxd # [0]
(wait for unit to become idle)
4. juju remove-storage fs/0
(should fail, saying that the storage is singular)
5. juju detach-storage fs/0
(should fail, saying that the storage is singular)
6. juju destroy-controller -y localhost --destroy-all-models

[0] The storagetest charm's metadata.yaml is as follows:
```yaml
name: storagetest
summary: ...
maintainer: Andrew Wilkins <andrew.wilkins@canonical.com>
description: ...
series: [xenial]
storage:
  fs:
    type: filesystem
```

## Documentation changes

None.

## Bug reference

None.
